Export JaCoCo coverage information (#359)
CodeIntelligenceTesting/jazzer: README.md

Particular stack traces can also be ignored based on their `DEDUP_TOKEN` by passing a comma-separated list of tokens
via `--ignore=<token_1>,<token2>`.

### Export coverage information

The internally gathered JaCoCo coverage information can be exported and used to generate coverage reports. These reports
can help identify areas that can not be reached through fuzzing and perhaps need changes to make them more accessible
for the fuzzer.

Similar to the JaCoCo `dump` command the flag `--coverage_dump=<file>` specifies the coverage export file, often called
`coverage.exec`, that should be generated after the fuzzing run. It contains a binary representation of the gathered
coverage data in the JaCoCo format.

The JaCoCo `report` command can be used to generate reports based on the generated coverage export. For example the
following command generates an HTML report in the folder `./report/` containing all classes available in `classes.jar`
and their coverage as captured in the export `coverage.exec`.
```shell

> java -jar jacococli.jar report coverage.exec \

Author: The Jazzer JaCoCo fork changes the method invocation handling. So I guess the probes could differ and result in incorrect coverage data. 

I included the JaCoCo CLI tool in our external build and updated the readme to show how the internal version could be used to generate a report. Nevertheless, this looks quite cumbersome and not the right place to build it to begin with. I tend to just point interested people to the fork. WDYT?

Particular stack traces can also be ignored based on their `DEDUP_TOKEN` by passing a comma-separated list of tokens
via `--ignore=<token_1>,<token2>`.

### Export coverage information

The internally gathered JaCoCo coverage information can be exported in human-readable and JaCoCo dump format.

> These can help identify code areas that can not be reached and perhaps need improvements to make them more accessible
> for the fuzzer.

Reviewer: I think it's important to offer clear guidance for this case. How about something like:

```suggestion
These can help identify code areas that have not been covered by the fuzzer and thus may require more comprehensive fuzz targets or a more extensive initial corpus to reach.
```

Author: done

The human-readable report contains coverage information, like branch and line coverage, on file level. It's useful to 
get a quick overview about the overall coverage. The flag `--coverage_report=<file>` can be used to generate it.

> Similar to the JaCoCo `dump` command the flag `--coverage_dump=<file>` specifies a coverage dump file, often called
> `coverage.exec`, that should be generated after the fuzzing run. It contains a binary representation of the gathered
> coverage data in the JaCoCo format.

Reviewer: ```suggestion
Similar to the JaCoCo `dump` command, the flag `--coverage_dump=<file>` specifies a coverage dump file, often called
`coverage.exec`, that is generated after the fuzzing run. It contains a binary representation of the gathered
coverage data in the JaCoCo format.
```

> 

Reviewer: I think we should provide a slightly more detailed description of what differs so that it's more likely people actually follow the suggestions. My attempt:

The JaCoCo `report` command can be used to generate reports based on this coverage dump. **Note:** The version
of the JaCoCo agent used by Jazzer internally differs slightly from the official one. As a result, a similarly modified version of the JaCoCo CLI tool has to be used to generate correct reports. The correct version is available at ....

The JaCoCo `report` command can be used to generate reports based on the coverage dump. The internally used version
differs slightly from the official one, which would lead to incorrect reports if used.

The internally used JaCoCo version is available on its [release page](https://github.com/CodeIntelligenceTesting/jacoco/releases)
as `zip` file. The report tool is located in the `lib` folder and can be used as described on the JaCoCo
[CLI documentation](https://www.eclemma.org/jacoco/trunk/doc/cli.html) page. For example the following command generates
an HTML report in the folder `report` containing all classes available in `classes.jar` and their coverage as

> captured in the export `coverage.exec`. Source code to color according to its coverage is searched for in

Reviewer: ```suggestion
captured in the export `coverage.exec`. Source code to include in the report is searched for in
```

Author: done

`some/path/to/sources`. After execution the `index.html` file in the output folder can be opened in a browser.
```shell
java -jar path/to/jacococli.jar report coverage.exec \
  --classfiles classes.jar \
  --sourcefiles some/path/to/sources \
  --html report \
  --name FuzzCoverageReport
```

Furthermore, it's also possible to directly use the CLI tools of the internal JaCoCo version via Bazel with the target

> `@jazzer_jacoco//:jacoco_cli`. The following command builds an HTML report similar to the mentioned one above:

Author: done

`coverage.exec`, that should be generated after the fuzzing run. It contains a binary representation of the gathered
coverage data in the JaCoCo format.

The JaCoCo `report` command can be used to generate reports based on the coverage dump. The internally used version

> differs slightly from the official one, and to prevent incorrect results it should be preferred.

Reviewer: IMO we should word this more strongly as incorrect coverage results aren't much better than no coverage result. ;-)

For example the following command generates an HTML report in the folder `/tmp/report/` containing all classes available in
`/classes.jar` and their coverage as captured in the export `/coverage.exec`.
```shell

> bazel run @jazzer_jacoco//:jacoco_cli -- \

Reviewer: I like that we offer a way to build this in-repo. Maybe we could just also add a link to the fork's release page to the preceding paragraph and provide instructions for both, with the fork's instructions coming first?

Author: Good idea, updated the readme.
